### PR TITLE
drivers: spi: spi_ll_stm32: SPI support for STM32MP1x SoC

### PR DESCRIPTION
**Pull request by yboutreux**
SPI support for STM32MP1x SoC. If SPI is selected, SPI4 (Arduino connector compatible SPI) and SPI5 will be enable by default on stm32mp157c_dk2 board.

PR related to reset flags commit : https://github.com/arnop2/zephyr_stm32mp1/pull/21

Signed-off-by: Yaël Boutreux <yael.boutreux@st.com>

### CODE REVIEW

**Comment by yboutreux** [comment]
@atorgue can you please review this PR ?

File: drivers/spi/spi_ll_stm32.c [hunk 1]
```diff
@@ -90,9 +99,20 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
...
+	while (!LL_SPI_IsActiveFlag_TXP(spi)) {
+		/* NOP if not enough space for a packet*/
+	}
+	if (LL_SPI_IsEnabled(spi) && LL_SPI_GetMode(spi)) {
+		LL_SPI_StartMasterTransfer(spi);
+	}
+	#else
 	while (!LL_SPI_IsActiveFlag_TXE(spi)) {
-		/* NOP */
+		/* NOP if buffer is not empty */
```
Comment (atorgue) [inline drivers/spi/spi_ll_stm32.c]: You modify no MP1 code

File: drivers/spi/spi_ll_stm32.c [hunk 2]
```diff
@@ -103,9 +123,18 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
...
+	while (!LL_SPI_IsActiveFlag_RXP(spi)) {
+		/* NOP if buffer has less than a packet to read*/
+	}
+	if (LL_SPI_IsEnabled(spi) && LL_SPI_GetMode(spi)) {
+		LL_SPI_StartMasterTransfer(spi);
+	}
+	#else
 	while (!LL_SPI_IsActiveFlag_RXNE(spi)) {
-		/* NOP */
+		/* NOP if buffer is empty */
```
Comment (atorgue) [inline drivers/spi/spi_ll_stm32.c]: ditto

File: drivers/spi/spi_ll_stm32.c [hunk 3]
```diff
@@ -125,6 +154,36 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
 /* Shift a SPI frame as slave. */
 static void spi_stm32_shift_s(SPI_TypeDef *spi, struct spi_stm32_data *data)
 {
+#ifdef CONFIG_SOC_SERIES_STM32MP1X
+	if (LL_SPI_IsActiveFlag_TXP(spi) && spi_context_tx_on(&data->ctx)) {
+		u16_t tx_frame = spi_stm32_next_tx(data);
+
+		if (SPI_WORD_SIZE_GET(data->ctx.config->operation) == 8) {
```
Comment (atorgue) [inline drivers/spi/spi_ll_stm32.c]: 8 could not replaced by a define ?

File: drivers/spi/spi_ll_stm32.c [hunk 4]
```diff
@@ -125,6 +154,36 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
...
+			LL_SPI_TransmitData8(spi, tx_frame);
+			spi_context_update_tx(&data->ctx, 1, 1);
```
Comment (atorgue) [inline drivers/spi/spi_ll_stm32.c]: avoid to use hard code value. Use define

File: drivers/spi/spi_ll_stm32.c [hunk 5]
```diff
@@ -139,7 +198,8 @@ static void spi_stm32_shift_s(SPI_TypeDef *spi, struct spi_stm32_data *data)
 		LL_SPI_DisableIT_TXE(spi);
 	}
 
-	if (LL_SPI_IsActiveFlag_RXNE(spi) && spi_context_rx_buf_on(&data->ctx)) {
+	if (LL_SPI_IsActiveFlag_RXNE(spi) &&
+	    spi_context_rx_buf_on(&data->ctx)) {
```
Comment (atorgue) [inline drivers/spi/spi_ll_stm32.c]: why do you change that ?

Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: "{" is exceeding the maximum "80 characters per line" rule, so I splitted the line

File: boards/arm/stm32mp157c_dk2/pinmux.c [hunk 6]
```diff
@@ -22,6 +22,20 @@ static const struct pin_config pinconf[] = {
 	{ STM32_PIN_PE7, STM32MP1X_PINMUX_FUNC_PE7_UART7_RX },
 	{ STM32_PIN_PE8, STM32MP1X_PINMUX_FUNC_PE8_UART7_TX },
 #endif  /* CONFIG_UART_7 */
+#ifdef CONFIG_SPI_4
+	{STM32_PIN_PE11, STM32MP1X_PINMUX_FUNC_PE11_SPI4_NSS},
+	{STM32_PIN_PE12, STM32MP1X_PINMUX_FUNC_PE12_SPI4_SCK},
+	{STM32_PIN_PE13, STM32MP1X_PINMUX_FUNC_PE13_SPI4_MISO |
+		STM32_OSPEEDR_VERY_HIGH_SPEED},
+	{STM32_PIN_PE14, STM32MP1X_PINMUX_FUNC_PE14_SPI4_MOSI},
+#endif /* CONFIG_SPI_5 */
```
[the remark this answers is not in the document]
Comment (yboutreux) [reply boards/arm/stm32mp157c_dk2/pinmux.c]: Indeed... Thank you !

File: boards/arm/stm32mp157c_dk2/pinmux.c [hunk 7]
```diff
@@ -22,6 +22,20 @@ static const struct pin_config pinconf[] = {
...
+#ifdef CONFIG_SPI_4
+	{STM32_PIN_PF6, STM32MP1X_PINMUX_FUNC_PF6_SPI5_NSS},
+	{STM32_PIN_PF7, STM32MP1X_PINMUX_FUNC_PF7_SPI5_SCK},
+	{STM32_PIN_PF8, STM32MP1X_PINMUX_FUNC_PF8_SPI5_MISO |
+		STM32_OSPEEDR_VERY_HIGH_SPEED},
+	{STM32_PIN_PF9, STM32MP1X_PINMUX_FUNC_PF9_SPI5_MOSI},
+#endif /* CONFIG_SPI_5 */
```
[the remark this answers is not in the document]
Comment (yboutreux) [reply boards/arm/stm32mp157c_dk2/pinmux.c]: Indeed... Thank you !

File: drivers/spi/spi_ll_stm32.c (hunk 5 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: This is now on a separate commit

File: drivers/spi/spi_ll_stm32.c (hunk 4 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: @atorgue can you look at _drivers: spi: spi_ll_stm32: explicit hard-coded values_ commit and tell me is this looks better to you ?

File: drivers/spi/spi_ll_stm32.c [hunk 8]
```diff
@@ -97,6 +97,10 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
 {
 	u16_t tx_frame;
 	u16_t rx_frame;
+	s8_t word_size = SPI_WORD_SIZE_GET(data->ctx.config->operation);
+	/* one chunk is 1 byte */
+	s8_t chunk_size = 1;
```
Comment (atorgue) [inline drivers/spi/spi_ll_stm32.c]: if it's a constant you don't need a variable but a define:

define ....

File: drivers/spi/spi_ll_stm32.c [hunk 9]
```diff
@@ -97,6 +97,10 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
 {
 	u16_t tx_frame;
 	u16_t rx_frame;
+	s8_t word_size = SPI_WORD_SIZE_GET(data->ctx.config->operation);
```
Comment (atorgue) [inline drivers/spi/spi_ll_stm32.c]: no. the change to do is not in this way. ou have to keep  SPI_WORD_SIZE_GET(data->ctx.config->operation) in the if statement

File: drivers/spi/spi_ll_stm32.c [hunk 10]
```diff
@@ -113,15 +117,14 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
 	}
 	#endif /* CONFIG_SOC_SERIES_STM32MP1X */
 
-	if (SPI_WORD_SIZE_GET(data->ctx.config->operation) == 8) {
+	if (word_size == 8) {
```
Comment (atorgue) [inline drivers/spi/spi_ll_stm32.c]: no

File: drivers/spi/spi_ll_stm32.c (hunk 9 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: @atorgue Modified to only define values, I still feel this is a bit confusing.
Though IMHO ` s8_t word_size = SPI_WORD_SIZE_GET(data->ctx.config->operation); ` was a good way to reduce the redundancy of ` spi_context_update_rx `usage (between 8bits and 16bits use case) without loosing code readability

Does this match what you were expecting ?

File: drivers/spi/spi_ll_stm32.c [hunk 11]
```diff
@@ -90,9 +99,20 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
 	u16_t rx_frame;
 
 	tx_frame = spi_stm32_next_tx(data);
+
+	#ifdef CONFIG_SOC_SERIES_STM32MP1X
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: `#ifdef`'s should not be indented

File: drivers/spi/spi_ll_stm32.c [hunk 12]
```diff
@@ -103,9 +123,18 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
 		spi_context_update_tx(&data->ctx, 2, 1);
 	}
 
+	#ifdef CONFIG_SOC_SERIES_STM32MP1X
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: ditto

File: drivers/spi/spi_ll_stm32.c [hunk 13]
```diff
@@ -125,6 +154,36 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
 /* Shift a SPI frame as slave. */
 static void spi_stm32_shift_s(SPI_TypeDef *spi, struct spi_stm32_data *data)
 {
+#ifdef CONFIG_SOC_SERIES_STM32MP1X
+	if (LL_SPI_IsActiveFlag_TXP(spi) && spi_context_tx_on(&data->ctx)) {
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: Can you add a short comment explaining why MP1 needs this specific code ?

File: drivers/spi/spi_ll_stm32.c [hunk 14]
```diff
@@ -303,12 +390,6 @@ static int spi_stm32_configure(struct device *dev,
 
 	LL_SPI_DisableCRC(spi);
 
-	if (config->operation & SPI_OP_MODE_SLAVE) {
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: If not MP1 specific, this should be moved in a dedicated commit

File: drivers/spi/spi_ll_stm32.c [hunk 15]
```diff
@@ -392,24 +498,45 @@ static int transceive(struct device *dev,
 
 #if defined(CONFIG_SPI_STM32_HAS_FIFO)
 	/* Flush RX buffer */
+#ifdef CONFIG_SOC_SERIES_STM32MP1X
+	while (LL_SPI_IsActiveFlag_RXP(spi)) {
+		(void)LL_SPI_ReceiveData8(spi);
+	}
+	spi_stm32_reset_errors(spi);
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c]: Do you think reset_error step can benefit to other series as well?

File: west.yml [hunk 16]
```diff
@@ -51,7 +51,7 @@ manifest:
       revision: b6c9262eed68000e69c92bef6e820cdbd5b0a32b
       path: modules/hal/st
     - name: hal_stm32
-      revision: b454ad819f4f10cb82ce5eb1d7f709a680bc61c5
```
Comment (erwango) [inline west.yml]: Remove this change as this PR is already merged.

Comment (yboutreux) [reply west.yml]: I am referring to this PR : https://github.com/arnop2/hal_stm32/pull/1
This will be updated once both PR will be on Zephyr GitHub

File: drivers/spi/spi_ll_stm32.c (hunk 15 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: This is quite odd, seems like in SPI HAL we reset errors during init, deinit and abort, this seems like a big mistake.
I am removing and testing it to see if SPI still works without it, nice catch !

Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: It works like a charm on SPI driver Zephyr tests and on my HCI-SPI project, removing it, thanks a lot again !

File: drivers/spi/spi_ll_stm32.c (hunk 14 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: This is a modification induced by the MP1, this was very very odd, not sure why but at the time during my testings I couldn't set the Master/Slave Mode until the NSS was set, I checked registers with devregs each time to see register state. (I don't know if @atorgue recall I found this issue at the beginning of April)
This is very odd since I couldn't see any indication about that in the SPI peripheral registers datasheet.
Retried I have the same behavior

File: drivers/spi/spi_ll_stm32.c (hunk 15 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: Edit : this is really useful and other board could benefit from IT. If everything goes well then we don't need it, but as soon a an error occur it will never be cleared, even after disabling and re-enabling SPI.
To clear then the only solution is to reboot the board.
This is why I reset the errors just before enabling the SPI (and after the RX buffer flush).
If you agree with this, I can add also the reset_error for other boards

File: west.yml (hunk 16 above)
Comment (erwango) [reply west.yml]: righ, txs

File: drivers/spi/spi_ll_stm32.c (hunk 14 above)
Comment (erwango) [reply drivers/spi/spi_ll_stm32.c]: Then, have you tested this change on other series?
So, even if is MP1 specific, since it impacts other series, I'll split in a previous commit, stating this is safe for other series and preparation work for MP1 support introduction.

File: drivers/spi/spi_ll_stm32.c (hunk 15 above)
Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: I am still confused to see there is no mechanism to reset coming errors.
It seems like "spi_stm32_complete" is acting like an abort mechanism (since this is called in spi_stm32_isr when an error is catched), maybe the reset_errors shouldn't be before the SPI_Enable but after the SPI_Disable inside the spi_stm32_complete ?
IMHO spi_stm32_complete which seems to act like an abort should implement SPI HAL function IRQHandler.
But for now since I only got MODF flag to clear, https://github.com/zephyrproject-rtos/zephyr/issues/17363 fix is enough instead of force clearing all flags

**Comment by yboutreux** [comment]
Updated the branch with the adding of a new commit which reset raised flags. This commit fixes the need to change NSS set and SPI Mode set order.
More information about this commit here : https://github.com/arnop2/zephyr_stm32mp1/pull/19

**Comment by yboutreux** [comment]
Updated, depends on https://github.com/arnop2/zephyr_stm32mp1/pull/21

File: drivers/spi/spi_ll_stm32.c [hunk 17]
```diff
@@ -177,23 +243,51 @@ static void spi_stm32_complete(struct spi_stm32_data *data, SPI_TypeDef *spi,
...
+#else
 	while (LL_SPI_IsActiveFlag_RXNE(spi)) {
 		(void) LL_SPI_ReceiveData8(spi);
 	}
-#endif
+#endif	/* CONFIG_SOC_SERIES_STM32MP1X */
+#endif	/* CONFIG_SPI_STM32_HAS_FIFO */
 
 	if (LL_SPI_GetMode(spi) == LL_SPI_MODE_MASTER) {
+#ifdef CONFIG_SOC_SERIES_STM32MP1X
```
Comment (yboutreux) [inline drivers/spi/spi_ll_stm32.c]: from @arnop2 :
"Please double check thes conditions. for me
LL_SPI_IsEnabled(spi) always true
!LL_SPI_IsActiveFlag_MODF(spi) always true
!LL_SPI_IsActiveFlag_TXC(spi) to check if always false as we are in spi_stm32_complete function..."

Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: For context, BSY flag isn't in the STM32MP1 SPI registers.
Following STM32L4 BSY Flag definition : 
```
/**
  * @brief  Get busy flag
  * @note   The BSY flag is cleared under any one of the following conditions:
  * -When the SPI is correctly disabled
  * -When a fault is detected in Master mode (MODF bit set to 1)
  * -In Master mode, when it finishes a data transmission and no new data is ready to be
  * sent
  * -In Slave mode, when the BSY flag is set to '0' for at least one SPI clock cycle between
  * each data transfer.
  * @rmtoll SR           BSY           LL_SPI_IsActiveFlag_BSY
  * @param  SPIx SPI Instance
  * @retval State of bit (1 or 0).
  */
```

Comment (yboutreux) [reply drivers/spi/spi_ll_stm32.c]: And it seems like LL_SPI_IsEnabled(spi) isn't always true, removing it from the the evaluation block lead the spi driver test to fail
